neqo-bin/src/server/mod.rs

```diff
@@ -188,28 +187,11 @@ impl Args {
     }
 }
 
-fn qns_read_response(filename: &str) -> Option<Vec<u8>> {
-    let mut file_path = PathBuf::from("/www");
-    file_path.push(filename.trim_matches(|p| p == '/'));
-
-    OpenOptions::new()
-        .read(true)
-        .open(&file_path)
-        .map_err(|_e| qerror!("Could not open {}", file_path.display()))
-        .ok()
-        .and_then(|mut f| {
-            let mut data = Vec::new();
-            match f.read_to_end(&mut data) {
-                Ok(sz) => {
-                    qinfo!("{} bytes read from {}", sz, file_path.display());
-                    Some(data)
-                }
-                Err(e) => {
-                    qerror!("Error reading data: {e:?}");
-                    None
-                }
-            }
-        })
+fn qns_read_response(filename: &str) -> Result<Vec<u8>, io::Error> {
+    let path: PathBuf = ["/www", filename.trim_matches(|p| p == '/')]
```

Comment:
This is not safe still.  We should be rejecting attempts to read files with a '/' or '\', but also paths that are equal to "." or ".." (though these a probably not good).  It's OK if this is not entirely robust, but still.

I would prefer to see a 404 response for files that don't exist.

Finally, "/www" should probably come in through the command line and default to "./www" (which might not exist, which is fine).

Reply:
Note that this function is used for the QUIC Interop Runner only. It is only executed when `--qns-test` is provided.

https://github.com/mxinden/neqo/blob/27ae7cbeb21d1858d3e371fa3a3ec01d029a93cd/neqo-bin/src/server/mod.rs#L343-L349

> Finally, "/www" should probably come in through the command line and default to "./www"

Given that it is only used through QUIC Interop, would this additional flag ever be used with anything other than `/www`?

> This is not safe still. We should be rejecting attempts to read files with a '/' or '', but also paths that are equal to "." or ".." (though these a probably not good). It's OK if this is not entirely robust, but still.

`..` and `.` will be sanitized by `url`, see following example:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=20a9aa366afa3df27f3f021d11e2d40d

> I would prefer to see a 404 response for files that don't exist.

:+1: added with https://github.com/mozilla/neqo/pull/1788/commits/cf0259124c6350eac3251b6f1a1c002f51000795.